datafusion/core/src/physical_optimizer/optimizer.rs

```diff
@@ -97,6 +97,8 @@ impl PhysicalOptimizer {
             // Note that one should always run this rule after running the EnforceDistribution rule
             // as the latter may break local sorting requirements.
             Arc::new(EnforceSorting::new()),
+            // TODO: `try_embed_to_hash_join` in the ProjectionPushdown rule would be block by the CoalesceBatches, so add it before CoalesceBatches. Maybe optimize it in the future.
```

Comment:
Could someone please file a ticket to track this work ? I can try and do so tomorrow if no one else has a chance

Reply:
Created: https://github.com/apache/arrow-datafusion/issues/9483